Migrate test runner to py.test

pferate opened the pull request:
Maintainer of pytest_gae needs to merge open PR and push to PyPI.
In the mean time, get code directly from GitHub mirror.

pferate committed:
Migrate test runner to py.test

Migrating test runner from `unittest2`/`nose` to `pytest`.
The pytest runner is also compatible with both unittest and nose tests.
Some of the benefits of PyTest include:
    * using plain asserts
    * function-based fixtures instead of setUp and tearDown
    * no strange camelCase methods

dhermes commented on oauth2client/crypt.py:
```diff
@@ -248,3 +237,25 @@ def verify_signed_jwt_with_certs(jwt, certs, audience=None):
+    define these on set up.
+    """
+    global Signer
+    global Verifier
```
👎  on the use of `global`

pferate replied:
They were global before, just being defined at the module level.  Pytest doesn't load module/global variables, which is why it was failing (unless someone knows of a way to do that).

I'll see if I can come up with a cleaner/better way to do that.

I came across [this answer](http://stackoverflow.com/a/7668273/398514) on StackOverflow regarding `__getattr__` at the module level, which referenced a hack that [Guido van Rossum suggested](https://mail.python.org/pipermail/python-ideas/2012-May/014969.html).

dhermes replied:
The issue isn't the use of globals / module level constants, just the use of the `global` keyword. It makes the module harder to understand.

pferate replied:
Guido's hack only works on Python 3, so that is not an option.

I was incorrect in my statement that Pytest didn't load module variables, modules are just reused between tests.  I've added some setUp/tearDown code to do some housecleaning when module variables are changed (at least for `oauth2client.crypt`)

@dhermes Thanks for making me take another look at things!

pferate commented:
@jonparrott Should we worry about populating the `conftest.py` file now?  Or should we wait until we start moving everything to fixtures (phase 3)?

pferate commented:
There is a bug in the 0.9 release of `flake8-import-order` (released yesterday) that is causing the `flake8` test to incorrectly fail.

I've already created an [issue](/PyCQA/flake8-import-order/issues/80) for it.

jonparrott commented:
@pferate if we need app engine stuff to work (which we do) we might as well populate conftest.py just for that. You only need a subset of the stuff in the example I linked.

pferate commented:
I copied the code from App Engine file you linked.  I moved the imports within `setup_sdk_imports` to within try/except blocks, so that the tests don't fail outside of App Engine.

I just added the fixture code for this PR to `pytest_configure()`.

These will be expanded more when we transition the tests here to fixtures.

theacodes commented on tests/conftest.py:
```diff
@@ -0,0 +1,149 @@
+        pass
+
+
+def import_appengine_config():
```
This probably isn't needed for oauth2client.

pferate replied:
Is this in reference to the `import_appengine_config` function or the try/except blocks above it?

theacodes replied:
the whole function.

theacodes commented on tests/conftest.py:
```diff
@@ -0,0 +1,149 @@
+    return tb
+
+
+def run_taskqueue_tasks(testbed, app):
```
This won't be needed here either.

pferate replied:
Similar question here. `run_taskqueue_tasks()` or `tb.` code?

theacodes replied:
this whole function

theacodes commented on tests/conftest.py:
```diff
@@ -0,0 +1,149 @@
+
+
+@pytest.fixture
+def run_tasks(testbed):
```
Won't be needed.

theacodes commented on tests/conftest.py:
```diff
@@ -0,0 +1,149 @@
+    util.positional_parameters_enforcement = util.POSITIONAL_EXCEPTION
+
+
+def pytest_runtest_call(item):
```
Can remove this.

theacodes commented on tests/conftest.py:
```diff
@@ -0,0 +1,149 @@
+def pytest_configure():
+    """conftest.py hook function for setting up SDK imports."""
+    setup_sdk_imports()
+    util.positional_parameters_enforcement = util.POSITIONAL_EXCEPTION
```
Add a comment here as to why this is needed?

theacodes commented on tox.ini:
```diff
@@ -18,74 +18,70 @@ deps = {[testenv]basedeps}
     coverage
-    nosegae
+    pytest-cov
+    # pytest_gae  # TODO: Move to official source when it is updated
```
With the conftest.py, pytest_gae won't be needed anymore.

pferate replied:
Where does the reference to `GAE_PYTHONPATH` get used?  Should I add a new fixture to update the `PATH` instead?

theacodes replied:
Probably just do it in setup_sdk_imports. (Alternatively other projects like python-docs-samples has nox/tox do it)

pferate replied:
I ended up updating `PYTHONPATH` in Tox for the `gae` environment.  I'm catching `ImportError` when running `setup_sdk_imports`, which may be a bit too generalized.  If `ImportError` is raised there, I'm assuming that it is not running within the `gae` test and those modules aren't needed.  If, for some reason, we are in the `gae` test and `ImportError` is raised, then the tests should fail.

jonparrott commented:
Please don't merge this until after the 3.0.0 release. I'll remove the label when it's safe.

jonparrott commented:
v3.0.0 is released so this can now be consider for merging.

pferate commented:
Should there be a 4.x or Dev branch, so that they can be kept separate from 3.x updates?

jonparrott commented:
> Should there be a 4.x or Dev branch, so that they can be kept separate from 3.x updates?

I don't think so, but @nathanielmanistaatgoogle is welcome to overrule that.

nathanielmanistaatgoogle commented on tests/conftest.py:
```diff
@@ -0,0 +1,106 @@
+
+    # Fixes timezone and other os-level items.
+    import google.appengine.tools.os_compat
+    (google.appengine.tools.os_compat)
```
I don't understand this at all?

theacodes replied:
Do you really want to understand App Engine magic?

Or do you mean the usage of `()` here? It just makes the linter ignore an unused import (we're importing for side-effects).

nathanielmanistaatgoogle replied:
Then is there a comment-level indication that can be used instead of a needless expression? `# pylint: disable=unused-import` or something perfectly clear like that?

theacodes replied:
This probably came from a project using flake8 instead of pylint, which has historically had pretty opaque error messages (E103, etc.).

pferate replied:
Removed `()` line and added a comment level flag to the import

nathanielmanistaatgoogle commented on tests/conftest.py:
```diff
@@ -0,0 +1,106 @@
+
+    # Setup the datastore and memcache stub.
+    # First, create an instance of the Testbed class.
+    tb = gaetestbed.Testbed()
```
Initialisms are discouraged for the same reason that single-letter identifiers are banned. Spell out something descriptive?

theacodes replied:
`testbed` would be a fine name here.

pferate replied:
Sounds good.  I just copied and pasted everything from GCP except for `util.positional_parameters_enforcement = util.POSITIONAL_EXCEPTION`.

nathanielmanistaatgoogle commented on tests/conftest.py:
```diff
@@ -0,0 +1,106 @@
+    (google.appengine.tools.os_compat)
+
+
+def setup_testbed():
```
"setup" is a noun; "set up" is a verb phrase. This should be `set_up_testbed`.

theacodes replied:
This might be slightly _too_ nitpicky. 😛

nathanielmanistaatgoogle commented:
Commit log message should be longer and describe why this change is being made.

nathanielmanistaatgoogle commented:
No dev branches. No branches other than `master` and `gh_pages`, actually. We're not that complicated a project.

pferate commented:
@jonparrott, I've been going over various ways of handling the GAE in this PR and I'd like some feedback.

Initially, I was using a fork of [pytest_gae](https://pypi.python.org/pypi/pytest_gae), due to the main library not working with newer versions of app engine.  Last week I got a PR merged to fix this and the library owner pushed it to PyPI.

You said to just use the code from the App Engine repo, but not all of it was utilized.  I also ran into an issue with the `cover` tox environment; one test expects GAE to be available and the other expects it to not be available.  If we had it always available, there is some trivial code that doesn't execute, because it's expecting the GAE library to not be there. There was no simple way of doing this without reproducing code from a plug-in.  Should I add a CLI option to load GAE (which would be added from plug-in)?  Also, we don't load the `google` module directly, so the line about reloading it always gets skipped.

I also came across another plug-in, (pytest-beds)[https://pypi.python.org/pypi/pytest-beds), which adds the following fixtures:

- testbed
- mailer
- channel
- urlfetch
- memcache
- taskqueue
- blobstore
- ndb
- users

`pytest-beds` seems like the most flexible option to me.

What are your thoughts on going forward with this?

jonparrott commented:
@pferate I'm going to send a PR to your fork to get the App Engine stuff where I want it.

Commit:
Re-configure App engine tests

jonparrott commented:
@pferate see https://github.com/pferate/oauth2client/pull/2

jonparrott commented:
I am now okay with this. @nathanielmanistaatgoogle to do the final review and merge.

(also there's a conflict, please fix)

pferate commented:
Rebased onto master.

Thanks for the help @jonparrott!

nathanielmanistaatgoogle commented on tests/contrib/appengine/conftest.py:
```diff
@@ -0,0 +1,61 @@
+    Return True to skip collection. Skips App Engine tests when --gae-sdk
+    is not specified.
+    """
+    if 'contrib/appengine' in str(path):
```
Does

```
return 'contrib/appengine' in str(path) and config.getoption('gae_sdk') is None
```

fit on one line?

pferate replied:
No.  If we do it that way, it would need to be split into 3 lines:

``` python
    return (
        'contrib/appengine' in str(path) and
        config.getoption('gae_sdk') is None)
```

We could take away the explicit `else`'s, but I know you prefer them.

``` python
    if 'contrib/appengine' in str(path):
        if config.getoption('gae_sdk') is None:
            return True
    return False
```

A one-line `if` statement is one character over the limit as well.

``` python
    if 'contrib/appengine' in str(path) and config.getoption('gae_sdk') is None:
```

Redefining `path` can get it all to fit in 2 lines:

``` python
    path = str(path)
    return 'contrib/appengine' in path and config.getoption('gae_sdk') is None
```

Not explicitly checking for `None` can get it into one line:

``` python
    return 'contrib/appengine' in str(path) and not config.getoption('gae_sdk')
```

nathanielmanistaatgoogle replied:
I like the three-line form best (it most clearly communicates "this is a simple conjuction") but could live with what's currently drafted. How much do you like the three-line form?

(One day an autoformatter will solve this!)

theacodes replied:
This is where we disagree - I don't like combining branching if statements like this unless they're directly related (e.g., `if x > 0 and x < screen.w`). But, I'll defer to you on this.

pferate replied:
Personally, I would go with:

``` python
    if 'contrib/appengine' in str(path):
        if config.getoption('gae_sdk') is None:
            return True
```

since, in this case, `True` is the only return value that does something special, all other return values are treated the same.

nathanielmanistaatgoogle replied:
I happen to think that `True` and `False` are equally special. Go with the conjunction.

pferate replied:
Will do.

nathanielmanistaatgoogle commented:
Looks mostly good; just that one comment.

pferate commented:
@jonparrott, with us moving the app engine tests into their own directory, should we move the app engine modules into their own directory as well?

Another coupled set of files are `contrib/gce` and `contrib/_metadata`.

jonparrott commented:
> with us moving the app engine tests into their own directory, should we move the app engine modules into their own directory as well?

I don't think so. There's no hard and fast rule that tests must be organized exactly as their module-under-test. I think defaulting to that but making exceptions where it makes testing easier (like we did here) makes sense. Practicality beats purity.

nathanielmanistaatgoogle commented on tests/contrib/appengine/conftest.py:
```diff
@@ -0,0 +1,61 @@
+        set_up_gae_environment(config.getoption('gae_sdk'))
+
+
+def pytest_ignore_collect(path, config):
```
Am I right to learn from there being no call sites of this function in added code that this, due to its name, is called by py.test? If so, maybe a better doc string for it would be something like `Directs py.test to skip App Engine tests when --gae-sdk is not specified.`?

pferate commented:
@nathanielmanistaatgoogle: PR updated, PTAL.

nathanielmanistaatgoogle commented on tests/contrib/appengine/conftest.py:
```diff
@@ -0,0 +1,57 @@
+
+def pytest_ignore_collect(path, config):
+    """Direct py.test to skip App Engine tests when --gae-sdk is not specified.
+
```
I think lines 51-53 can be dropped; a one-line summary is now fine for this function. "Return True to skip collection" probably belongs in the abstract "super" function, wherever that's defined.

pferate replied:
Dropped.  I had to reword it because the `"""` made it go over 80 characters.

nathanielmanistaatgoogle commented:
One last tweak; sorry.

pferate commented:
PR updated.  @nathanielmanistaatgoogle PTAL.

nathanielmanistaatgoogle commented:
What's with that test failure?

Is it desired that this go in as two commits? (No pushing in either direction; just want to confirm.)

pferate commented:
Looks like it may be related to #590.  This is what is in my Travis logs:
```
# Travis ships with an old version of PyPy, so install at least version 2.6.
if [[ "${TOX_ENV}" == "pypy" ]]; then
    git clone https://github.com/yyuu/pyenv.git ${HOME}/.pyenv
    ${HOME}/.pyenv/bin/pyenv install pypy-2.6.0
fi
fatal: destination path '/home/travis/.pyenv' already exists and is not an empty directory.
```
In my repo, `pypy` passed the first time, but failed when I reran the test.

As for the commits, one was a merged PR from @jonparrott helping me sort out some of the GAE tests, so I didn't want to lose his work within mine.  Functionally, they could be squashed.

jonparrott commented:
The pypy failure is my fault. We probably just need to check for the non-existence of {$HOME}/.pyenv before doing the git clone.

We can do it in this PR or I can send a separate one.

nathanielmanistaatgoogle commented:
Separate pull request with the fix in isolation, please.

jonparrott commented:
@pferate It's fixed, rebase and try again?

pferate commented:
@jonparrott Rebased and waiting on Travis.

One weird thing that I've been seeing recently is `py33` tests have been [failing in my fork on Travis](https://travis-ci.org/pferate/oauth2client/jobs/150778885), but have been [passing on the main repo](https://travis-ci.org/google/oauth2client/jobs/150778904).  They also pass when I run them on my local system.

The 2 linked Travis jobs are from the same commit from yesterday.  It looks like my fork is collecting the django tests (it shouldn't) and the main repo correctly skips them.  Before I dig much further, do you have any idea of what's going on?

As long as it passes on the main repo, it doesn't really matter; but this seems strange.

nathanielmanistaatgoogle commented:
@pferate: are you waiting for answers from @jonparrott before this gets merged?

(I am completely fine with the pull request content and metadata; feel free to merge when conversation is again concluded.)

pferate commented:
No need to wait.  I just had some weird, unique test results that I'm trying to figure out on my end.
